dev/mcal/src/Adc.cpp

```diff
@@ -0,0 +1,185 @@
+/**
+ * @file Adc.cpp
+ * @brief 
+ * @version 0.1
+ * @date 2024-07-20
+ * 
+ * @copyright Copyright (c) 2024
+ * 
+ */
+#include "mcal/inc/stm32f103xx.h"
+#include "Types.h"
+#include "Assert.h"
+#include "BitManipulation.h"
+#include "Rcc.h"
+#include "Adc.h"
+
+using namespace stm32::utils::bit_manipulation;
+using namespace stm32::registers::rcc; 
+using namespace stm32::dev::mcal::adc;
+using namespace stm32::registers::adc;
+
+ASSERT_STRUCT_SIZE(ADCRegDef, (sizeof(RegWidth_t) * 20));
+
+ASSERT_MEMBER_OFFSET(ADCRegDef, SR,          0);
+ASSERT_MEMBER_OFFSET(ADCRegDef, CR1,          sizeof(RegWidth_t) * 1);
+ASSERT_MEMBER_OFFSET(ADCRegDef, CR2,         sizeof(RegWidth_t) * 2);
+ASSERT_MEMBER_OFFSET(ADCRegDef, SMPR1,         sizeof(RegWidth_t) * 3);
+ASSERT_MEMBER_OFFSET(ADCRegDef, SMPR2,         sizeof(RegWidth_t) * 4);
+ASSERT_MEMBER_OFFSET(ADCRegDef, JOFR1,         sizeof(RegWidth_t) * 5);
+ASSERT_MEMBER_OFFSET(ADCRegDef, JOFR2,        sizeof(RegWidth_t) * 6);
+ASSERT_MEMBER_OFFSET(ADCRegDef, JOFR3,        sizeof(RegWidth_t) * 7);
+ASSERT_MEMBER_OFFSET(ADCRegDef, JOFR4,        sizeof(RegWidth_t) * 8);
+ASSERT_MEMBER_OFFSET(ADCRegDef, HTR,        sizeof(RegWidth_t) * 9);
+ASSERT_MEMBER_OFFSET(ADCRegDef, LTR,        sizeof(RegWidth_t) * 10);
+ASSERT_MEMBER_OFFSET(ADCRegDef, SQR1,        sizeof(RegWidth_t) * 11);
+ASSERT_MEMBER_OFFSET(ADCRegDef, SQR2,        sizeof(RegWidth_t) * 12);
+ASSERT_MEMBER_OFFSET(ADCRegDef, SQR3,        sizeof(RegWidth_t) * 13);
+ASSERT_MEMBER_OFFSET(ADCRegDef, JSQR,        sizeof(RegWidth_t) * 14);
+ASSERT_MEMBER_OFFSET(ADCRegDef, JDR1,        sizeof(RegWidth_t) * 15);
+ASSERT_MEMBER_OFFSET(ADCRegDef, JDR2,        sizeof(RegWidth_t) * 16);
+ASSERT_MEMBER_OFFSET(ADCRegDef, JDR3,        sizeof(RegWidth_t) * 17);
+ASSERT_MEMBER_OFFSET(ADCRegDef, JDR4,        sizeof(RegWidth_t) * 18);
+ASSERT_MEMBER_OFFSET(ADCRegDef, DR,        sizeof(RegWidth_t) * 19);
+
+ADC::ADC(const ADCConfig& config) : config_(config) {
+    switch (config_.number) {
+        case kADC1 : ADC_reg = (reinterpret_cast<volatile ADCRegDef*>(ADC1));
+                     break;
+        case kADC2 : ADC_reg = (reinterpret_cast<volatile ADCRegDef*>(ADC2));
+                     break;
+        default    : break;
+    }
+}
+
+void ADC::Init() {
+    ADC_reg->CR2.registerVal = 0;  // Reset ADC
+    ADC_reg->CR2.ALIGN = static_cast<RegWidth_t>(config_.alignment);
+    if (config_.trigSource != kSoftware) {
+        ADC_reg->CR2.ADON = 1;
+        return;
+    } else if (config_.mode == kInjected) {
+        ADC_reg->CR2.JEXTTRIG = 1;
+        ADC_reg->CR2.JEXTSEL = static_cast<RegWidth_t>(config_.trigSource);
+    } else {
+        ADC_reg->CR2.EXTTRIG = 1;
+        ADC_reg->CR2.EXTSEL = static_cast<RegWidth_t>(config_.trigSource);
+    }
+    ADC_reg->CR2.ADON = 1;
```

Comment:
**Note** if the `config_.trigSource` is the **ADC** will start the conversion  immediately because the `ADON` bit will be set again and the second set will start the conversion.